runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -303,14 +309,34 @@ public void processElements(Iterable<WindowedValue<InputT>> values) throws Excep
       windowsToConsider.addAll(processElement(windowToMergeResult, value));
     }
 
-    // Trigger output from any window for which the trigger is ready
-    for (W mergedWindow : windowsToConsider) {
-      ReduceFn<K, InputT, OutputT, W>.Context directContext =
-          contextFactory.base(mergedWindow, StateStyle.DIRECT);
-      ReduceFn<K, InputT, OutputT, W>.Context renamedContext =
-          contextFactory.base(mergedWindow, StateStyle.RENAMED);
-      triggerRunner.prefetchShouldFire(mergedWindow, directContext.state());
-      emitIfAppropriate(directContext, renamedContext);
+    if (!windowsToConsider.isEmpty()) {
```

Comment:
Just a drive-by comment, but are we sure reiteration is cheap here? This appears to be a per-element cost, and given the scale we've found that introducing even small extra per-element cost can be quite noticeable.